src/utils.js

```diff
@@ -200,32 +220,183 @@ export const getAgeData = (birthMonth, birthYear, age, ageMonths) => {
+      ),
+    );
+
+    const similarity = maxIdenticalResponse / responses.length;
+
+    // Calculate response similarity based on maxIdenticalResponse
+    if (similarity >= RESPONSE_SIMILARITY_THRESHOLD) {
+      flags.push('responsesTooSimilar');
+    }
+
+    // Calculate accuracy based on the number of correct responses
+    const numCorrect = correct?.filter((x) => x === 1).length ?? 0;
+    if (numCorrect / correct.length <= ACCURACY_THRESHOLD) {
+      flags.push('accuracyTooLow');
+    }
+
+    const isReliable = flags.filter((x) => !ignoredReliabilityFlags.includes(x)).length === 0;
+
+    return { flags, isReliable };
+  };
+}
+
+/**
+ * Tracks response times and invokes callback function when response times
+ * exceed specified thresholds.
+ *
+ * @class ResponseTimeTracker
+ * @param {Function} evaluateValidity function to be called to generate an decision on the
+ *    reliability of a run
+ * @param {Function} addEngagementFlags function passed through to update the run's
+ *  firekit object with flags tripped
+ * @param {number} minResponsesRequired The minimum number of responses required before
+ *   checking for threshold exceedance.
+ * @property {Array<number>} _responseTimes An array to store the response times.
+ * @property {Array<number>} _responses An array to store the responses 
+ * (keypresses or button choices).
+ * @property {Array<number>} _correct An array to store the correctness of the responses.
+ */
+export class ValidityEvaluator {
+  constructor({
+    evaluateValidity = () => {},
+    addEngagementFlags = () => {},
+    minResponsesRequired = 0,
+  }) {
+    this.evaluateValidity = evaluateValidity;
+    this.addEngagementFlags = addEngagementFlags;
+    this.minResponsesRequired = minResponsesRequired;
+    this._responseTimes = [];
+    this._responses = [];
+    this._correct = [];
+  }
+
+  /**
+   *  @function startNewBlockValidation Called when a new block is started to reset the
+   * data arrays and update the evaluateValidity function if needed
+   * @param {Function} evaluateValidity
+   */
+  startNewBlockValidation(evaluateValidity = this.evaluateValidity) {
+    this.evaluateValidity = evaluateValidity;
+    this._responseTimes = [];
+    this._responses = [];
+    this._correct = [];
+  }
+
+  /**
+   * Updates ValidityEvaluator arrays (responseTimes, response, and accuracy) with
+   *
+   * @function addResponseData
+   * @param {number} responseTime Time it took for a user to respond to a stimulus
+   * @param {string} response  Choice that a user responded with
+   * ex: left_arrow, right_arrow, button_3
+   * @param {number} isCorrect 1 if a user answered correctly, 0if answered incorrectly
+   */
+  addResponseData(responseTime, response, isCorrect) {
+    // TODO: make response data arrays private
+    this._responseTimes.push(responseTime);
+    this._responses.push(response);
+    this._correct.push(isCorrect);
+
+    if (this._responseTimes.length >= this.minResponsesRequired) {
```

Comment:
i am bit confused where you will call the baseEvaluateValidity function. through this.evaluateValidity? if yes, why do you still need line 394 to 399? and where can I see isReliable is default to be false?

Reply:
The `baseEvaluateFunction` would be called on line 388 as you mentioned. (It is passed into the class through the callback function `evaluateValidity` -- You can see an example of this in utils.spec.js) Since the baseEvaluateFunction only returns the flags that should be attached to the run, but does not actually set the flag's values in the DB, lines 394-399 are needed to update a run's flags and reliability value using the passed in `addEngagementFlags` function. 

The default value of isReliable is set to false [here](https://github.com/yeatmanlab/roar-firekit/pull/78/files)